onnxruntime/test/contrib_ops/tokenizer_test.cc

```diff
@@ -0,0 +1,607 @@
+  }
+  // Test of the overlapping search patterns
+  // The spec mandates that the patterns that appear
+  // in the separators earlier must be matched first.
+  {
+    // In this case the first pattern must match first
+    // and there would be no match for the second
+    std::vector<std::string> separators = {
+        u8"Абсу",
+        u8"су"};
+
+    OpTester test("Tokenizer", opset_ver, domain);
+    InitTestAttr(test, false, separators, 1);
+
+    std::vector<int64_t> dims{1};
+    std::vector<std::string> input{u8"Абсу中文"};
+    test.AddInput<std::string>("T", dims, input);
+
+    std::vector<int64_t> output_dims(dims);
+    // Must drop the beginning of the word that
+    // also contains the second separator
+    output_dims.push_back(int64_t(1));
+    std::vector<std::string> output{u8"中文"};
+
+    test.AddOutput<std::string>("Y", output_dims, output);
+
+    test.Run(OpTester::ExpectResult::kExpectSuccess);
+  }
+  // Test of the overlapping search patterns
+  // The spec mandates that the patterns that appear
+  // in the separators earlier must be matched first.
+  {
+    // In this case the first pattern must match first
+    // and there would be no match for the second
+    std::vector<std::string> separators = {
+        u8"Абсу",
+        u8"су"};
+
+    OpTester test("Tokenizer", opset_ver, domain);
+    InitTestAttr(test, false, separators, 1);
+
+    std::vector<int64_t> dims{1};
+    std::vector<std::string> input{u8"Абсусусу中文"};
+    test.AddInput<std::string>("T", dims, input);
+
+    std::vector<int64_t> output_dims(dims);
+    // Must drop the beginning of the word that
+    // also contains the second separator
+    output_dims.push_back(int64_t(1));
+    std::vector<std::string> output{u8"中文"};
+
+    test.AddOutput<std::string>("Y", output_dims, output);
+
+    test.Run(OpTester::ExpectResult::kExpectSuccess);
+  }
+  // Test of the overlapping search patterns
+  // The spec mandates that the patterns that appear
+  // in the separators earlier must be matched first.
+  {
+    // In this case the first pattern must match first
+    // and there are more than one overlapping matches for the first
+    // so the earlier match for the first wins.
+    // and there would be no match for the second
+    std::vector<std::string> separators = {
+        u8"усу",
+        u8"Абсу"};
+
+    OpTester test("Tokenizer", opset_ver, domain);
+    InitTestAttr(test, false, separators, 1);
+
+    std::vector<int64_t> dims{1};
+    std::vector<std::string> input{u8"Абсусусу中文"};
+    test.AddInput<std::string>("T", dims, input);
+
+    std::vector<int64_t> output_dims(dims);
+    // Must drop the beginning of the word that
+    // also contains the second separator
+    output_dims.push_back(int64_t(2));
+    std::vector<std::string> output{u8"Абс", u8"су中文"};
+
+    test.AddOutput<std::string>("Y", output_dims, output);
+
+    test.Run(OpTester::ExpectResult::kExpectSuccess);
+  }
+}
+}  // namespace test
```

Comment:
I added the following unit test, it happens when a separator is a prefix of another one. I don't know if this case is expected but if not we should raised an exception or warning in the constructor.

TEST(ContribOpTest, TokenizerWithSeparators__) {
  // Separators and strings with a mix of latin, Spanish, Cyrillic and Chinese
  // characters and with start/end text markers
  // [C] dimensions
  // Output [C][D]
  {
    std::vector<std::string> separators = {
        u8";",
        u8";;;"};

    OpTester test("Tokenizer", opset_ver, domain);
    InitTestAttr(test, true, separators, 1);

    std::vector<int64_t> dims{4};
    std::vector<std::string> input{u8"a;b", u8"a;;;b", u8"b;c;;;d;e", u8"a;;b;;;c"};
    test.AddInput<std::string>("T", dims, input);

    std::vector<int64_t> output_dims(dims);
    // Must split both in 2
    output_dims.push_back(int64_t(6));
    std::vector<std::string> output {
        start_mark, u8"a", u8"b", end_mark, padval, padval,
        start_mark, u8"a", u8"b", end_mark, padval, padval, 
        start_mark, u8"b", u8"c", u8"d", u8"e", end_mark,
        start_mark, u8"a", /* u8"", if we allow null string*/ u8"b", u8"c", end_mark, padval,
    };

    test.AddOutput<std::string>("Y", output_dims, output);

    std::cout << "*****\n";
    test.Run(OpTester::ExpectResult::kExpectSuccess);
  }  // namespace test
}

And it returns the following:
---
a;b
---
a
b
---
b;c
d;e
---
a;;b
c

Reply:
Thank you for the test. I will investigate. The spec allows separators that are prefixes of each other. This is what it says: "If 'separators' is a L-element array, there will be L rounds of tokenization using one stop word. More specifically, in the first round, the first element in 'separators' is used to tokenize each string in the input. Then, the second element in 'separators' will be used to tokenize the resulted strings produced at the first round."

---
In reply to: [239077639](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r239077639) [](ancestors = 239077639)